lib/kastlex/offsets_cache.ex

```diff
@@ -24,80 +26,97 @@ defmodule Kastlex.OffsetsCache do
-    :erlang.send_after(state.refresh_timeout_ms, Kernel.self(), @refresh)
+  def handle_info({@refresh, id, generation}, %{generation: generation} = state) do
+    refresh_leader_offsets(id, generation, state.refresh_timeout_ms)
+    {:noreply, state}
+  end
+  def handle_info({@refresh, _id, _generation}, state) do
+    # wrong generation, drop the message
     {:noreply, state}
   end
 
-  def handle_info(msg, state) do
-    Logger.error "Unexpected msg: #{inspect msg}"
+  def handle_info(@refresh, state) do
+    leaders = Kastlex.MetadataCache.get_leader_ids()
+    state = case Enum.sort(leaders) == state.cached_leaders do
+              true -> state
+              false ->
+                Logger.info "resetting requests and leaders cache. " <>
+                  "Leaders = #{inspect leaders}, " <>
+                  "cached leaders = #{inspect state.cached_leaders}"
+                # bump generation
+                generation = state.generation + 1
+                Enum.each(leaders,fn(id) ->
+                  refresh_leader_offsets(id, generation, state.refresh_timeout_ms) end)
+                %{state | :cached_leaders => Enum.sort(leaders), :generation => generation}
+            end
     {:noreply, state}
   end
 
   def terminate(reason, _state) do
-    Logger.info "#{inspect Kernel.self} is terminating: #{inspect reason}"
+    Logger.info "terminating: #{inspect reason}"
   end
 
-  def refresh_leader_offsets(client_config, {id, topics}) do
-    leader = Kastlex.MetadataCache.get_brokers |> Enum.find(fn(x) -> x.id == id end)
-    case leader do
-      nil ->
-        Logger.error("Error refresing offsets for node #{id}: not found in broker metadata")
-      _ ->
-        refresh_leader_offsets(client_config, leader, topics)
-    end
+  def refresh_leader_offsets(id, generation, refresh_timeout_ms) do
+    parent = self()
+    Kernel.spawn(
+      fn ->
+        lps = Kastlex.MetadataCache.get_leader_partitions(id)
+        topics = Enum.reduce(lps, [],
+          fn({t, partitions}, acc) ->
+            partition_fields = Enum.reduce(partitions, [],
+            fn(p, acc2) ->
+              [[partition: p,
+                timestamp: -1,
+                max_num_offsets: 1] | acc2]
+            end)
+            [[topic: t, partitions: partition_fields] | acc]
+          end)
+        client_id = Kastlex.get_brod_client_id()
+        leader = Kastlex.MetadataCache.get_leader(id)
+        {:ok, conn} = :brod_client.get_connection(client_id, String.to_charlist(leader.host), leader.port)
+        vsn = :brod_kafka_apis.pick_version(conn, :list_offsets)
+        request_fields = [replica_id: -1, topics: topics]
+        request = :kpro.make_request(:list_offsets, vsn, request_fields)
+        {:ok, kpro_rsp(msg: msg)} = :kpro.request_sync(conn, request, 10_000)
+        handle_offsets_response(msg[:responses])
+        :erlang.send_after(refresh_timeout_ms, parent, {@refresh, id, generation})
```

Comment:
we need a big `try catch` for the spawned `fn`, and only do this `send_after` in `after` clause.

Reply:
@zmstone fixed